src/function/execute.rs

```diff
@@ -360,3 +544,44 @@ impl<C: Configuration> Drop for ClearCycleHeadIfPanicking<'_, C> {
         }
     }
 }
+
+/// Returns the key of any potential outer cycle head or `None` if there is no outer cycle.
+///
+/// That is, any query that's currently blocked on the result computed by this query (claiming it results in a cycle).
+fn outer_cycle(
+    zalsa: &Zalsa,
+    zalsa_local: &ZalsaLocal,
+    cycle_heads: &CycleHeads,
+    current_key: DatabaseKeyIndex,
+) -> Option<DatabaseKeyIndex> {
+    // First, look for the outer most cycle head on the same thread.
+    // Using the outer most over the inner most should reduce the need
+    // for transitive transfers.
+    // SAFETY: We don't call into with_query_stack recursively
+    if let Some(same_thread) = unsafe {
+        zalsa_local.with_query_stack_unchecked(|stack| {
+            stack
+                .iter()
+                .find(|active_query| {
+                    cycle_heads.contains(&active_query.database_key_index)
+                        && active_query.database_key_index != current_key
```

Comment:
Is this enough to identify strongly connected components? Do we never need to prefer the outermost query for deeply nested cycles?